### PR TITLE
chore: install via ephemeral go module rather than tools module

### PR DESCRIPTION
**Pull request by devigned**
**What this PR does / why we need it**:

Tools used in the build have been know to have version conflicts between them. As of now, we have been putting all of our tools into the `hack/tool/go.mod`. This has worked relatively well, but it could be better.

This PR provides an isolated way of installing bins by creating a tmp dir and creating a short lived module in the tmp dir. The install is done in an isolated manor and the bin is placed in the GOBIN dir.

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -33,11 +33,11 @@ export GO111MODULE=on
 # Directories.
 ROOT_DIR:=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))
 TOOLS_DIR := hack/tools
-TOOLS_BIN_DIR := $(TOOLS_DIR)/bin
+TOOLS_BIN_DIR := $(abspath $(TOOLS_DIR)/bin)
```
Comment (ncdc) [inline Makefile:36]: macOS does not have `abspath` out of the box - if you need this, I would do `cd $(TOOLS_DIR)/bin && pwd -P`

File: Makefile [hunk 2]
```diff
@@ -33,11 +33,11 @@ export GO111MODULE=on
...
 TEMPLATES_DIR := $(ROOT_DIR)/templates
-BIN_DIR := bin
-RELEASE_NOTES_BIN := bin/release-notes
+BIN_DIR := $(abspath $(ROOT_DIR)/bin)
```
Comment (ncdc) [inline Makefile:38]: Ditto re macOS/`abspath`

File: scripts/go_install.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
+#!/usr/bin/env bash
+# Copyright 2018 The Kubernetes Authors.
```
Comment (ncdc) [inline scripts/go_install.sh:2]: 2020, or was this copied from another repo?

File: Makefile (hunk 1 above)
Comment (devigned) [reply Makefile:36]: I actually built this on OS X. 

```
$ make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```

File: scripts/go_install.sh (hunk 3 above)
Comment (devigned) [reply scripts/go_install.sh:2]: copypasta from https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/scripts/kind-with-registry.sh, which may have been copypasta from elsewhere.

I thought it more appropriate to have the actual year of inclusion vs. inaccurate. 

I'm happy to change it. What should it be?

**Comment by devigned** [comment]
@ncdc, I had run into issues with kustomize requiring a singular ref of client.go. 

This is a singular issue which doesn't tell the whole story. This is not first time installing modules within a tools directory has been needlessly painful. There is no reason each tool should have a dependency on a dependency version of another tool. All we care about is that we have a bin tool locked to a given version, so we can have repeatable builds.

In a perfect world, we, the language community, would solve this problem in the Go tool. Until that happens, this seems to be the most pragmatic approach to solving the problem.

If installing bins this way offends others' sensibilities, I'm happy change to a community acceptable solution.

File: Makefile (hunk 1 above)
Comment (ncdc) [reply Makefile:36]: My bad - was thinking this was the shell abspath, not the make built-in.

File: scripts/go_install.sh (hunk 3 above)
Comment (ncdc) [reply scripts/go_install.sh:2]: That script has 2020 though?

If the content of this file is net new, we should do 2020.

**Comment by ncdc** [comment]
Thanks for the explanation. No offense taken 😄!

File: scripts/go_install.sh (hunk 3 above)
Comment (devigned) [reply scripts/go_install.sh:2]: Oops, you’re right. I’ll fix that. Thx!

Comment (devigned) [reply scripts/go_install.sh]: Fixed

**Comment by vincepri** [comment]
Thanks for doing this, we should do this upstream as well at some point, we ran into the same kustomize issue and decided to punt it until later

File: controllers/azuremachine_reconciler.go [hunk 4]
```diff
@@ -346,6 +346,7 @@ func (s *azureMachineService) createVirtualMachine(nicName string) (*infrav1.VM,
 func GetControlPlaneMachines(machineList *clusterv1.MachineList) []*clusterv1.Machine {
 	var cpm []*clusterv1.Machine
 	for _, m := range machineList.Items {
+		m := m
```
Comment (CecileRobertMichon) [inline controllers/azuremachine_reconciler.go:349]: what is this for? is it related to the change or is it an unrelated fix sneaking into the PR?

Comment (devigned) [reply controllers/azuremachine_reconciler.go:349]: It's related since golangci-lint won't pass unless it's fixed. Outside of that, I would have broken into another commit.

Comment (devigned) [reply controllers/azuremachine_reconciler.go:349]: The for loop is passing a reference to a var. If that var is not captured in the closure of the for loop, then bad things can happen.

Comment (devigned) [reply controllers/azuremachine_reconciler.go:349]:
```
controllers/azuremachine_reconciler.go:349:33: G601: Implicit memory aliasing in for loop. (gosec)
		if util.IsControlPlaneMachine(&m) {
		                              ^
```

**Comment by CecileRobertMichon** [comment]
/hold

I'm getting:

```
 cerobert@DESKTOP-HMFMKI4  ~/go/src/sigs.k8s.io/cluster-api-provider-azure   remotes/devigned/go-install ● ⍟17  rm -rf hack/tools/bin                                       SIGINT(2) ↵  10591  11:54:24
 cerobert@DESKTOP-HMFMKI4  ~/go/src/sigs.k8s.io/cluster-api-provider-azure   remotes/devigned/go-install ● ⍟17  make generate                                                         ✔  10592  11:54:26
/Library/Developer/CommandLineTools/usr/bin/make generate-go
GOBIN=/Users/cerobert/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin ./scripts/go_install.sh sigs.k8s.io/controller-tools/cmd/controller-gen@v0.2.9
go: creating new go.mod: module fake/mod
go: finding sigs.k8s.io/controller-tools/cmd v0.2.9
go: finding sigs.k8s.io/controller-tools/cmd/controller-gen v0.2.9
go: finding sigs.k8s.io v0.2.9
GOBIN=/Users/cerobert/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin ./scripts/go_install.sh github.com/golang/mock/mockgen@v1.4.3
go: creating new go.mod: module fake/mod
go: finding github.com v1.4.3
go: finding github.com/golang v1.4.3
go: finding github.com/golang/mock/mockgen v1.4.3
GOBIN=/Users/cerobert/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin ./scripts/go_install.sh k8s.io/code-generator/cmd/conversion-gen@v0.18.2
go: creating new go.mod: module fake/mod
go: finding k8s.io/code-generator/cmd/conversion-gen v0.18.2
go: finding k8s.io/code-generator/cmd v0.18.2
go: finding k8s.io v0.18.2
/Users/cerobert/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/controller-gen \
		paths=./api/... \
		paths=./exp/api/... \
		object:headerFile=./hack/boilerplate/boilerplate.generatego.txt
-: api/v1alpha2/zz_generated.conversion.go:1:1: expected 'package', found 'EOF'
/Users/cerobert/go/src/sigs.k8s.io/cluster-api-provider-azure/api/v1alpha2/azurecluster_conversion.go:20:16: no such package located
262:23: missing ',' in argument list
265:1: missing ',' in argument list
266:9: missing ',' in composite literal
267:6: expected '==', found '='
268:7: expected '==', found '='
269:2: missing ',' before newline in argument list
270:1: expected operand, found 'if'
271:9: missing ',' in composite literal
272:6: expected '==', found '='
273:7: expected '==', found '='
274:2: missing ',' before newline in argument list
Error: not all generators ran successfully
run `controller-gen paths=./api/... paths=./exp/api/... object:headerFile=./hack/boilerplate/boilerplate.generatego.txt -w` to see all available markers, or `controller-gen paths=./api/... paths=./exp/api/... object:headerFile=./hack/boilerplate/boilerplate.generatego.txt -h` for usage
make[1]: *** [generate-go] Error 1
make: *** [generate] Error 2
```

from this branch checked out

**Comment by devigned** [comment]
@CecileRobertMichon that's weird... Seems to be working in CI. What's going on?

**Comment by CecileRobertMichon** [comment]
/hold cancel 

I re-ran it after clearing the git diff and now it's working, not sure what happened there
